### ISSUE TITLE
#179 prevents concurrent function invocation

### ISSUE THREAD

**Issue by urp** [crossbario/autobahn-cpp#183]
I recently updated autobahn-cpp in my embedded project and discovered a regression intoduced with #179 which breaks my use-case.
My application uses the bonefish router and autobahn-cpp connected via a [native_transport](https://github.com/urp/bonefish-native) running on the same io_service. 
Since I provide several long running operations for multiple clients I want to make sure multiple invocations are able to run in parallel on the same thread. I found a kind of elegant solution using boost::fibers [integrated with boost::asio](https://www.boost.org/doc/libs/1_66_0/libs/fiber/doc/html/fiber/integration/deeper_dive_into___boost_asio__.html) via the same io_service used for the wamp router and client. 
Unfortunately the changes introduced with 2325b7a  make it impossible to return from a function handler without sending an empty result marking the wamp_invocation as not sendable(). 
This problem should affect all attempts to handle invocations in parallel, since this requires that ONLY the handler is responsible for the wamp_invocation after it was passed to the handler.

Another way to fix #178 without causing these problems would be to call empty_result() from the destructor of the wamp_invocation (if it is still sendable()). 

Here is the code I use to handle function invocations:
```c++
session->provide
( fn_name
, [fn]( autobahn::wamp_invocation wi )
  {
    boost::fibers::fiber
    ( boost::fibers::launch::dispatch
    , [ fn = std::move(fn), wi]{ fn( wi ); }
    ).detach();
  }
);
```

**Comment by mdegel** [comment]
This is a valid point. It seems like somewhat of on oversight on my behalf, when I was dealing with this my use case then.

After further looking into your issue I would say it doesn't only break your use case with running multiple  invocations in one thread. Instead it generally prevents one from storing the invocation and returning results any time later.

From what I can tell after looking into your suggested solution, this approach seems to be feasible. For the invocation it should be irrelevant which thread will call the destructor therefore sending the potentially empty result. 
The only issue I can see so far is that one could potentially "leak" the invocation, by permanently storing it. This could lead to the earlier issues which resulted in https://github.com/crossbario/autobahn-cpp/pull/179. However this would be an explicit programming error and not a potential oversight like simply not calling `empty_result()` at the end of a function.

@oberstet: Is there anything from your perspective which would speak against the proposed solution from @urp?

**Comment by oberstet** [comment]
@mdegel 

> it generally prevents one from storing the invocation and returning results any time later.

yep, the issue is broader in scope. I also overlooked that .. we need more unit and integration tests!

@urp 

> call empty_result() from the destructor of the wamp_invocation (if it is still sendable())

that sounds good! as far as I can see, it would permit your use case (multiple fibers, one thread), as well as the postponed processing of invocation (storing the invocation somewhere in the invoked procedure, and hence the destructor won't fire ..).
